### ISSUE TITLE
How do I use the local Singularity image file

### ISSUE THREAD

**Issue by kevin-furant** [broadinstitute/cromwell#6685]
Hi,
    I have a local Singularity mirror file named qc.sif. I did not find any examples of configuring the local .sif file into the cromwell.example.conf file in the documentation you gave me. I checked some configurations and failed. I would like you to give an example of the configuration and a short runtime section of the.wDL file. Thank you very much!

**Comment by kevin-furant** [comment]
Call-cache cannot be used if a Singularity mirror is used. So Singularity becomes useless in the WDL process, where both call-cache and container are needed.

**Comment by conanyangqun** [comment]
I use this config in a SGE backend for singularity.

```
Singularity {
            actor-factory = "cromwell.backend.impl.sfs.config.ConfigBackendLifecycleActorFactory"

            config {
                concurrent-job-limit = 100
                exit-code-timeout-seconds = 120
                runtime-attributes = """
                String sif
                Float? memory_gb
                String? bind_path
                """

                submit = """
                qsub \
                -terse \
                -V \
                -b y \
                -N ${job_name} \
                -wd ${cwd} \
                -o ${out}.qsub \
                -e ${err}.qsub \
                ${"-l mem_free=" + memory_gb + "g"} \
                singularity exec -e --bind ${cwd}:${cwd} \
                ${"--bind " + bind_path} \
                ${sif} \
                ${job_shell} ${script}
                """

                job-id-regex = "(\\d+)"
                check-alive = "qstat -j ${job_id}"
                kill = "qdel ${job_id}"
            }
        }
```
Every task should have a `String sif`, point to the path to sif file. You can modify this according to your need.

**Comment by olivierlabayle** [comment]
Hi, has this been addressed somehow? I am working on a cluster with restricted internet access and no docker registry. The above requires to modify the workflow tasks definition. Is there another way to point cromwell to a local singularity sif file? At the moment Cromwell complains the image has an invalid syntax and that it can't mirror it.

**Comment by natechols** [comment]
@olivierlabayle This is definitely possible but requires some creative approaches to the `submit-docker` command in your `cromwell.conf` - the previous comment is hinting at the correct solution but it's a little messier.  For example, this is how we handled it for running Cromwell in PacBio's SMRT Link server:
```
        SINGULARITY_CONTAINER_ARG="docker://${docker}"
        SINGULARITY_IMAGE_FILE=$(echo "${docker}" | sed 's#.*/##; s/:/_/g; s/$/.sif/;')
        if [ -f "$MY_SINGULARITY_IMAGES/$SINGULARITY_IMAGE_FILE" ]; then
            echo "Using cached image file $SINGULARITY_IMAGE_FILE"
            SINGULARITY_CONTAINER_ARG="$MY_SINGULARITY_IMAGES/$SINGULARITY_IMAGE_FILE"
        fi
```
And the eventual submission command (our own shell script wrapping sbatch/qsub/etc.) is running this:
```
singularity -v --nocolor exec $SINGULARITY_ARGS --bind ${cwd}:${docker_cwd} $SINGULARITY_CONTAINER_ARG /usr/bin/env bash ${docker_script}"
```
To summarize: we are translating the `docker` image specified in the WDL to a local .sif path, and if that SIF exists we run singularity with that directly, if not we pull from the repo.  So if you pre-populate your local cache with all of the SIFs you need for your workflow, it should execute correctly.  This has apparently worked for our customers on servers with similarly restricted access (and fixed the instability that occurs when we try to pull containers on the fly during workflow execution).

**Comment by olivierlabayle** [comment]
Thank you @natechols , what should the `docker` variable point to? If I keep the original docker image name, it leads to "Error connecting to https://auth.docker.io" because there is no connection and no local docker registry. This error seems to be thrown before the execution of the script.

**Comment by natechols** [comment]
`docker` is coming from Cromwell, it's defined earlier in the backend settings:
```
        runtime-attributes = """
            Int cpu = 1
            Int gpu = 0
            Int? memory_mb
            String username = "smrtlink"
            String? docker
        """
```
and the actual string in the WDL is something like this:
```
runtime {
  docker: "google/deepvariant:1.6.1"
  ...
}
```
So that means I need to have `deepvariant_1.6.1.sif` in my cache directory.

**Comment by olivierlabayle** [comment]
I haven't been able to make this work. Cromwell seems to be trying to establish a connection with docker.io before the submission script, and effectively, can never reach it (no internet connection). I suspect this might work for you if your HPC has a access to the internet. The only way I could make this work was to use a solution similar to below, that is disable the `submit-docker` and hardcode the singularity image to be used in the submission script (instead of declaring a `sif` in every task):

> I use this config in a SGE backend for singularity.
> 
> ```
> Singularity {
>             actor-factory = "cromwell.backend.impl.sfs.config.ConfigBackendLifecycleActorFactory"
> 
>             config {
>                 concurrent-job-limit = 100
>                 exit-code-timeout-seconds = 120
>                 runtime-attributes = """
>                 String sif
>                 Float? memory_gb
>                 String? bind_path
>                 """
> 
>                 submit = """
>                 qsub \
>                 -terse \
>                 -V \
>                 -b y \
>                 -N ${job_name} \
>                 -wd ${cwd} \
>                 -o ${out}.qsub \
>                 -e ${err}.qsub \
>                 ${"-l mem_free=" + memory_gb + "g"} \
>                 singularity exec -e --bind ${cwd}:${cwd} \
>                 ${"--bind " + bind_path} \
>                 ${sif} \
>                 ${job_shell} ${script}
>                 """
> 
>                 job-id-regex = "(\\d+)"
>                 check-alive = "qstat -j ${job_id}"
>                 kill = "qdel ${job_id}"
>             }
>         }
> ```
> 
> Every task should have a `String sif`, point to the path to sif file. You can modify this according to your need.
